### DISCUSSION TITLE
Get a unfrozen object from "snapshot"

### DISCUSSION THREAD

**Discussion by saivanji** [pmndrs/valtio#701] [General]
`snapshot` function returns a frozen object with `writable: false` in it's descriptor.

Is it possible to get an object which will be writable from the `snapshot` function?

**Comment by dai-shi** [comment]
You have to copy the snapshot object, for example using `_.cloneDeep`.

If you feel it's not optimal (I do so actually), there might be a better way. Why not using proxy `state` instead?
Or, deep cloning is unavoidable for certain use cases, like we do in `proxyWithHistory`.
https://github.com/pmndrs/valtio/blob/7fe26c2a8ff2df8a5b8df019f8878dbf48e40876/src/vanilla/utils/proxyWithHistory.ts#L10

**Reply by saivanji** [reply]
@dai-shi yeah, deep cloning is undesirable (we use it inside the `onChange` handler, which will hurt typing experience).

Can you please clarify, what you mean by "using proxy state instead"?

**Reply by dai-shi** [reply]
"proxy state" means an object returned by `proxy()`. So, it simply means "can you avoid relying on snapshots?"

Because snapshot is a frozen snapshot by nature, you can't modify.
```js
const originalObj = { count: 0 };
const snapshotObj = Object.freeze({ ...originalObj });
++originalObj.count; // shouldn't change the snapshot object
// If you need to modify the snapshot object, you need to copy:
const newModifiableObj = { ...snapshotObj };
++newModifiableObj.count; // now it's possible to modify
```
